File: HomelabAndroid/app/src/main/java/com/homelab/app/data/local/SiteStore.kt
package com.homelab.app.data.local

import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import com.homelab.app.domain.model.Site
import com.homelab.app.domain.model.SiteRegistry
import com.homelab.app.domain.model.Tenant
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map
import kotlinx.serialization.json.Json

/**
 * Persists the device-local [SiteRegistry] in the same Preferences DataStore [TenantStore] uses,
 * under its own JSON key. All mutation rules live in [SiteRegistry]; this class only stores the
 * result and generates ids for new sites.
 *
 * A device that never creates a site keeps the default empty [SiteRegistry.INITIAL] and never
 * writes the key.
 */
@Singleton
class SiteStore @Inject constructor(
    private val dataStore: DataStore<Preferences>,
    private val json: Json
) {
    private val key = stringPreferencesKey(STORE_KEY)

    val registry: Flow<SiteRegistry> = dataStore.data.map { decode(it[key]) }

    suspend fun current(): SiteRegistry = decode(dataStore.data.first()[key])

    /** Creates a site for [tenantRef] with a fresh id and returns the stored registry. */
    suspend fun addSite(tenantRef: String, name: String): SiteRegistry {
        val site = Site(
            id = "site-${UUID.randomUUID()}",
            tenantRef = Tenant.refOrDefault(tenantRef),
            name = name.trim()
        )
        return update { it.adding(site) }
    }

    suspend fun renameSite(id: String, name: String): SiteRegistry =
        update { it.renaming(id, name.trim()) }

    suspend fun removeSite(id: String): SiteRegistry = update { it.removing(id) }

    private suspend fun update(transform: (SiteRegistry) -> SiteRegistry): SiteRegistry {
        var result = SiteRegistry.INITIAL
        dataStore.edit { preferences ->
            result = transform(decode(preferences[key])).normalized()
            preferences[key] = json.encodeToString(SiteRegistry.serializer(), result)
        }
        return result
    }

    private fun decode(raw: String?): SiteRegistry {
        if (raw.isNullOrBlank()) return SiteRegistry.INITIAL
        return runCatching {
            json.decodeFromString(SiteRegistry.serializer(), raw).normalized()
        }.getOrDefault(SiteRegistry.INITIAL)
    }

    private companion object {
        const val STORE_KEY = "site_registry_v1"
    }
}

File: HomelabSwift/Homelab/Stores/SiteStore.swift
import Foundation
import Observation

/// Persists the device-local `SiteRegistry` as a single JSON blob in `UserDefaults`, under its own
/// key alongside `TenantStore`. All mutation rules live in `SiteRegistry`; this type only stores
/// the result and generates ids for new sites.
///
/// A device that never creates a site keeps `SiteRegistry.initial` and never writes the key.
@Observable
@MainActor
final class SiteStore {

    private(set) var registry: SiteRegistry

    @ObservationIgnored private let defaults: UserDefaults
    private static let storageKey = "site_registry_v1"

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        self.registry = Self.snapshot(defaults: defaults)
    }

    nonisolated static func snapshot(defaults: UserDefaults = .standard) -> SiteRegistry {
        guard
            let data = defaults.data(forKey: storageKey),
            let decoded = try? JSONDecoder().decode(SiteRegistry.self, from: data)
        else {
            return .initial
        }
        return decoded.normalized()
    }

    /// Creates a site for `tenantRef` with a fresh id and returns the stored registry.
    @discardableResult
    func addSite(tenantRef: String, name: String) -> SiteRegistry {
        let site = Site(
            id: "site-\(UUID().uuidString.lowercased())",
            tenantRef: Tenant.refOrDefault(tenantRef),
            name: name.trimmingCharacters(in: .whitespacesAndNewlines)
        )
        return apply { $0.adding(site) }
    }

    @discardableResult
    func renameSite(id: String, name: String) -> SiteRegistry {
        apply { $0.renaming(id: id, to: name.trimmingCharacters(in: .whitespacesAndNewlines)) }
    }

    @discardableResult
    func removeSite(id: String) -> SiteRegistry {
        apply { $0.removing(id: id) }
    }

    private func apply(_ transform: (SiteRegistry) -> SiteRegistry) -> SiteRegistry {
        let next = transform(registry).normalized()
        registry = next
        persist(next)
        return next
    }

    private func persist(_ registry: SiteRegistry) {
        guard let data = try? JSONEncoder().encode(registry) else { return }
        defaults.set(data, forKey: Self.storageKey)
    }
}
